### ISSUE TITLE
Doesn't work for emacs installed through `brew cask install emacs`

### ISSUE THREAD

**Issue by hyiltiz** [cask/cask#446]
EmacsWiki says `brew cask install emacs` is the recommended way for the installing emacs for macOS. I installed `cask` using the `curl` command shown in `README.md`. But `cask` doesn't seem to have detected the presence of `emacs` under `/Applications/Emacs.app/`.

**Comment by Fuco1** [comment]
Different cask: https://sourabhbajaj.com/mac-setup/Homebrew/Cask.html

**Comment by Fuco1** [comment]
Thanks for the clarification!  I can't help you much as I don't use the Apple platform, but hopefully someone will.

I suppose everything works fine when you just type `emacs` in the terminal? Maybe `homebrew-cask` does not add a symlink to somewhere on your path.  Would that solve the problem (temporarily, it's not a great solution)?

**Comment by hyiltiz** [comment]
Yes, it does open `emacs` since the symlinks are created and configured properly.

**Comment by greened** [comment]
I just ran into this problem.  It's pretty much a show-stopper for folks on OSX.  Is there some way for Cask to find the emacs on the PATH?

**Comment by conao3** [comment]
Hi, recently I join this project and I notice this issue via your comment.

I don't have macOS, so I don't familiar with mac and brew.

If you install Emacs via brew, you can invoke Emacs via `emacs` in your terminal? (in another phrase, what output when you exec `which emacs`)

**Comment by hyiltiz** [comment]
I installed emacs via `brew cask install emacs`, which ultimately produces `/Application/Emacs.app/`; I then manually create the symlinks and PATH by doing `sudo ln -s /Application/Emacs.app/Contents/MacOS/Emacs /usr/local/bin/Emacs` (only the Emacs GUI, not terminal). If you do `brew install emacs`, it is different (as described above).

**Comment by conao3** [comment]
I've rent mac and try it but it works well...?

1. `brew install cask`

   ![Capture 2020-12-09 8 10 49](https://user-images.githubusercontent.com/4703128/101553041-50c93100-39f7-11eb-8045-2e7183e868d1.png)

2. manual instlal Cask with `brew cask install emacs`

   ![Capture 2020-12-09 8 16 15](https://user-images.githubusercontent.com/4703128/101553090-68081e80-39f7-11eb-957e-e29ec7039282.png)

And it seems your environment is the same as me (you can find emacs when `which emacs`)
Are there still problems?

**Comment by hyiltiz** [comment]
Hmm, that seems to resolve it; I guess at some point that `homebrew-cask` recipe started to `Linking Binary 'Emacs' to '/usr/local/bin/emacs`. Thanks!

**Comment by srcreigh** [comment]
I landed here today via google with the same issue.

I apparently had an extra emacs install under brew --cask. I had to run `brew uninstall --cask emacs` as well as `brew uninstall emacs`. I also deleted Emacs.app from `/Applications`. Following that, `brew install cask` worked.

Thanks for the encouragement in this thread especially for @conao3 who rented a Mac just to test for us.

**Comment by srcreigh** [comment]
Actually I think this issue should still be open.

There are two ways of installing emacs with brew. The `brew install --cask emacs` formula is the only one that supports a GUI. The emacs which gets installed when you run `brew install cask` doesn't give you any GUI. It confirms as such on the emacs wiki [here](https://wikemacs.org/wiki/Installing_Emacs_on_OS_X). And if you have the GUI formula installed already, `brew install cask` fails.

```
~/Code % brew install --cask emacs
Updating Homebrew...
==> Auto-updated Homebrew!
Updated 1 tap (homebrew/core).
==> Updated Formulae
Updated 2 formulae.

==> Downloading https://emacsformacosx.com/emacs-builds/Emacs-27.1-1-universal.dmg
==> Downloading from https://emacsformacosx.com/download/emacs-builds/Emacs-27.1-1-universal.dmg
######################################################################## 100.0%
==> Installing Cask emacs
==> Moving App 'Emacs.app' to '/Applications/Emacs.app'
==> Linking Binary 'Emacs' to '/usr/local/bin/emacs'
==> Linking Binary 'ebrowse' to '/usr/local/bin/ebrowse'
==> Linking Binary 'emacsclient' to '/usr/local/bin/emacsclient'
==> Linking Binary 'etags' to '/usr/local/bin/etags'
🍺  emacs was successfully installed!
~/Code % brew install cask
==> Downloading https://homebrew.bintray.com/bottles/emacs-27.1.catalina.bottle.1.tar.gz
Already downloaded: /Users/shane/Library/Caches/Homebrew/downloads/6f24b416d3f6dbe84688a2ebe27abf49118d8bc7f46e13bb9b4728b843f2f00d--emacs-27.1.catalina.bottle.1.tar.gz
==> Downloading https://github.com/cask/cask/archive/v0.8.5.tar.gz
Already downloaded: /Users/shane/Library/Caches/Homebrew/downloads/c6258144ae7417dd614623a5655cdfc627f57ab790a25305757b214b1bdd4b43--cask-0.8.5.tar.gz
==> Installing dependencies for cask: emacs
==> Installing cask dependency: emacs
==> Pouring emacs-27.1.catalina.bottle.1.tar.gz
==> emacs cask is installed, skipping link.
==> Caveats
To have launchd start emacs now and restart at login:
  brew services start emacs
Or, if you don't want/need a background service you can just run:
  emacs
==> Summary
🍺  /usr/local/Cellar/emacs/27.1: 4,009 files, 104.1MB
==> Installing cask
Error: An exception occurred within a child process:
  RuntimeError: /usr/local/opt/emacs not present or broken
Please reinstall emacs. Sorry :(
~/Code %
```

TL;DR the `brew install cask` formula requires a silly version of emacs that doesn't even come with a GUI.. it'd be nice to be able to install cask without limiting yourself to TUI emacs.

**Comment by conao3** [comment]
OK, but I don't maintenance the cask formula, I don't know where the formula is.
It is a good way to contact the formula maintainer to fix its recipe.
(I apologize in advance if the cask project managed it.
I am now the maintainer of this project, but I don't have a mac, and the management of the formulas is clearly without a maintainer.)

**Comment by srcreigh** [comment]
Thanks again @conao3. I'm happy to help out. For background I am trying to work on the tree-sitter-emacs library, there are instructions which involve setting up cask in the readme over [here](https://github.com/ubolonton/emacs-tree-sitter)

The cask formula is on Github [here](https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/cask.rb) I found it by searching on [this Homebrew site](https://formulae.brew.sh/formula/cask)

I did find a way to solve this problem. I had an idea that the line `emacs cask is installed, skipping link.` while trying to install cask was significant, because it complained about a missing link later: `/usr/local/opt/emacs not present or broken`. So I tried re-linking emacs with the command `brew link --overwrite emacs` and it worked.

Not only does this allow me to install cask with the GUI emacs formula. but also I tried uninstalling everything and starting over, and surprisingly I didn't have to run `brew link --overwrite emacs` again. I could just run `brew install --cask emacs` and then `brew install cask` and it worked. I am using a 7 year old MBP, so it's possible I had some broken symlinks that are independent of the installation.

I am very new to this cask tool. I tested that the installation works by running `cask --help`. I also ran the `./bin/setup` command from emacs-tree-sitter, which calls out to the cask command to install things.

TL;DR:
This issue seems to be caused by installation-independent broken symlinks

To fix on my computer, I ran these steps

1. Uninstall all versions of emacs. `brew uninstall emacs` `brew uninstall --cask emacs`
2. Install the Homebrew Cask version which has a GUI and Emacs.app: `brew install --cask emacs`
3. Tried installing cask via `brew install cask`
3. Due to the error about broken symlinks, I set up links again using `brew link --overwrite emacs`
4. Install cask via `brew install cask` which now works.

**Comment by conao3** [comment]
Thanks. Then, finally, what we need to do? Clean macOS and `brew` environment?

**Comment by hyiltiz** [comment]
@conao3 Didn't srceigh's post above yours help?

**Comment by conao3** [comment]
I want to just confirm. If we have a clean macOS and `brew` environment, we can install `cask` via `brew install cask` and it works well?

**Comment by hyiltiz** [comment]
If you did `brew install --cask emacs` according to srcreigh; arguable if you installed your emacs in other ways but configured your paths well, it could also work.

**Comment by caisah** [comment]
I think the docs should state that the required file should be based on the location of `brew` files.

Right now it's `(require 'cask "/usr/local/share/emacs/site-lisp/cask/cask.el")` but if 
`brew --cellar cask` returns `/opt/homebrew/Cellar/cask` 

the correct path will be `/opt/homebrew/share/emacs/site-lisp/cask/cask.el`.

**Comment by dickmao** [comment]
The installation method has changed.

Ideally just one method is supported for all POSIX-like systems, and preferably one the maintainers have control over (and isn't a bot-bumped homebrew formula that last saw a substantive change six years ago).

**Comment by lmBored** [comment]
Wait I'm so confused. So am I supposed to do ```
Brew install cask
``` after ```
Brew install --cask emacs
```? Because as on gnu emacs site, I just have to install the "--cask emacs"

**Comment by jcs090218** [comment]
If you want to use Cask, you should do `brew install cask`. So you are right. :)

However, I don't recommend people to use Cask anymore. There is another tool I've created, [Eask](https://github.com/emacs-eask/cli). You can treat it as a successor to Cask, and it's so much more powerful now!

To install it:

```sh
$ brew install node
$ npm install -g @emacs-eask/cli
```

**Comment by dickmao** [comment]
> If you want to use Cask, you should do brew install cask. So you are right. :)

This is incorrect.  Refer to the README on how to install.  If you don't have GNU make, then you're out of luck.

> [Eask] so much more powerful now!

As near as I can tell, Eask is junk but probably less fussy than Cask.

**Comment by lmBored** [comment]
So if I do `brew install --cask emacs` as instructed on https://www.gnu.org/software/emacs/download.html#nonfree then I'm good to go right? (I'm on macos ventura)

**Comment by lmBored** [comment]
Thanks a lot for helping!!

**Comment by jcs090218** [comment]
> This is incorrect. Refer to the README on how to install. If you don't have GNU make, then you're out of luck.

Wait, so I am correct? He's not using "this" Cask, and just want the Emacs to be installed. TBH, the name Cask (this project) and brew cask is very confusing. 😅 

> As near as I can tell, Eask is junk but probably less fussy than Cask.

It would be great if you were willing to point out the reason why instead of calling it junk. But I'm also not too surprised since you are doing this everywhere (GNU Emacs mailing list, Emacs China, etc). So I won't care too much about your opinions since it would probably be meaningless. 🤔

**Comment by dickmao** [comment]
> The slogan for subversion was "cvs done right," and if you start with
> that slogan there's nowhere you can go.  There is no way you can do
> cvs right. -- Linus Torvalds

Replace subversion and cvs with Eask and Cask respectively.

Cask was ill-conceived from the start because it added its own opaque
layer of command line options, as if emacs's built-in set weren't
already painful enough.  To compound the folly it saddled repos with
an incongruous "Cask" file.

Eask, its clumsy name notwithstanding, doubles down on Cask's
mis-architecture, adding to the sprawl, preserving the obtuse "Cask"
file, and generally increasing cognitive demand.  And in a possibly
tongue-in-cheek stroke of self-caricature, it wraps the whole
enterprise in nodejs.

What you want in a sandboxing library is minimalism.  You want to be
able to invoke it transparently with "emacs -Q --batch -l
package-development-library..."

> It would be great if you willing to point out the reason

I trust you found my explanation "great."  I did not enjoy writing it,
so you owe me one.

**Comment by jcs090218** [comment]
Just to let you know, I simply ignoring it since I don't find it useful. 😅 Also I am not going to reply more (like other people did). Feel free to discuss Eask's issues on the Eask repo so we don't spam unrelated messages here.

But thanks for sharing thoughts! ;)
